controllers/apps/transformer_component_deletion.go

```diff
@@ -36,42 +54,131 @@ func (t *componentDeletionTransformer) Transform(ctx graph.TransformContext, dag
 		return nil
 	}
 
+	reqCtx := intctrlutil.RequestCtx{
+		Ctx:      transCtx.Context,
+		Log:      transCtx.Logger,
+		Recorder: transCtx.EventRecorder,
+	}
 	graphCli, _ := transCtx.Client.(model.GraphClient)
 	comp := transCtx.Component
+	cluster, err := t.getCluster(transCtx, comp)
+	if err != nil {
+		return newRequeueError(requeueDuration, err.Error())
+	}
+
+	// step1: update the component status to deleting
+	if comp.Status.Phase != appsv1alpha1.DeletingClusterCompPhase {
+		comp.Status.Phase = appsv1alpha1.DeletingClusterCompPhase
+		graphCli.Status(dag, comp, transCtx.Component)
+		return newRequeueError(time.Second*1, "updating component status to deleting")
+	}
+
+	// step2: do the pre-terminate action if needed
+	if err := component.ReconcileCompPreTerminate(reqCtx, t.Client, cluster, comp, dag); err != nil {
+		if intctrlutil.IsTargetError(err, intctrlutil.ErrorWaitCacheRefresh) {
+			return newRequeueError(time.Second*3, "wait for preTerminate to be done")
+		}
+		return err
+	}
+
+	// step3: delete the sub-resources
+	compShortName, err := component.ShortName(cluster.Name, comp.Name)
+	if err != nil {
+		return err
+	}
+	var (
+		toDeleteKinds   []client.ObjectList
+		toPreserveKinds []client.ObjectList
+	)
+	// by default, we inherit cluster termination policy to control the sub-resources deletion
```

Comment:
> It is not reasonable to inherit the termination policy from cluster directly. Consider a specific issue: the termination policy exhibits a certain degree of atomicity at the cluster level. However, at the component level, it's not possible to guarantee the synchronization between multiple components because their deletions are triggered by the user. This may potentially result in inconsistent behavior among the components, unless significant complex works are carried out to resolve the underlying consistency problems.

By default, all resources under comp are deleted. To be compatible with the behavior and semantics when a cluster's TerminationPolicy is set to Halt upon deletion, some resources are preserved in this scenario.